Fixed #27719 -- Added QuerySet.alias() to allow creating reusable aliases.

alextatarinov opened the pull request:
Added queryset.alias() method to allow creating an alias for the expression
without including the expression in the select clause and later reuse.
Using .alias() will produce the same side effects (joins and group by) as .annotate() immediately,
to avoid different query results depending on the alias used later.
The allowed use of an alias is limited intentionally to methods that can use the expression
without including it into the select clause.

airstrike commented:
Per the chain on django-developers, it sounds like this is somewhat related to #12602, so perhaps @claudep and / or @charettes would be interested. I'm just dying to have this rolled out and am wondering what I can do to contribute, which is probably very little given my knowledge of the ORM internals is quite limited.

charettes commented on docs/ref/models/querysets.txt:
```diff
@@ -268,6 +268,51 @@ control the name of the annotation::
+    >>> from django.db.models import Count
+    >>> blogs = Blog.objects.alias(entries=Count('entry')).filter(entries__gt=5)
+
+Despite the value not being selected, all other effects of using ``alias()`` are
+the same as of ``annotate()``.
+
+``alias()`` can be used in conjunction with
+:meth:`~django.db.models.query.QuerySet.annotate()`,
+:meth:`~django.db.models.query.QuerySet.filter()`,
+:meth:`~django.db.models.query.QuerySet.order_by()`
+and :meth:`~django.db.models.query.QuerySet.update()`.
+To use aliased expression with other methods, one must promote alias to annotation:
+
+    >>> from django.db.models import Count, F
+    >>> blogs = Blog.objects.alias(entries=Count('entry')).annotate(entries=F('entries'))
+
```
Should the `aggregate` limitation be explicitly specified here (talking about `Cannot aggregate over the alias. Use annotate() to promote the alias`) or is it implied by the fact it's not mentioned.

I find

> Despite the value not being selected, all other effects of using ``alias()`` are the same as of ``annotate()``.

contradicts

> ... To use aliased expression with other methods, one must promote alias to annotation:

Maybe we should drop the _Despite_ paragraph?

alextatarinov replied:
`aggregate` limitation is implied by the fact it's not mentioned in the list of supported methods. We can probably add it as an example `with other methods (e.g. aggregate) `

I think we can drop the first  the Despite paragraph.

charettes commented on tests/annotations/tests.py:
```diff
@@ -120,17 +120,62 @@ def test_basic_annotation(self):
 
+    def test_basic_alias(self):
+        books = Book.objects.alias(
+            is_book=Value(1, output_field=IntegerField())
```
No need for `Value` wrapping since 1e38f1191de21b6e96736f58df57dfb851a28c1f

```suggestion
            is_book=1,
```

Ditto for all `Values` uses below.

alextatarinov replied:
Without `Value` there is an error `TypeError: QuerySet.annotate() received non-expression(s): 1.`. Am I missing something?

charettes replied:
Ahh looks like you'll need to keep passing `Value` in this case but you can drop the `output_field`.
```suggestion
            is_book=Value(1)
```

charettes commented on tests/annotations/tests.py:
```diff
@@ -120,17 +120,62 @@ def test_basic_annotation(self):
+        books = Book.objects.alias(
+            is_book=Value(1, output_field=IntegerField())
+        )
+        self.assertNotIn('is_book', str(books.query))
```
What about

```suggestion
        self.assertNotIn('is_book', books.values().first())
```

charettes commented on tests/annotations/tests.py:
```diff
@@ -120,17 +120,62 @@ def test_basic_annotation(self):
+        books = Book.objects.alias(
+            is_book_alias=Value(1, output_field=IntegerField())
+        ).annotate(is_book=F('is_book_alias'))
+        self.assertNotIn('is_book_alias', str(books.query))
```
Ditto I'd avoid asserting against the string representation of the query, and assert against the returned `.values()`.

alextatarinov replied:
@charettes `.values()` changes quite a lot in a query generation. Is it safe to assert against it?

charettes replied:
I you don't want to use `.values()` you could simply use `not hasattr(book, 'is_book_alias`)`?

charettes commented on tests/annotations/tests.py:
```diff
@@ -120,17 +120,62 @@ def test_basic_annotation(self):
+        books = Book.objects.alias(
+            is_book_alias=Value(1, output_field=IntegerField())
+        ).annotate(is_book=Coalesce('is_book_alias', 0))
+        self.assertNotIn('is_book_alias', str(books.query))
```
Ditto

charettes commented on docs/ref/models/querysets.txt:
```diff
@@ -268,6 +268,51 @@ control the name of the annotation::
+``alias()``
+~~~~~~~~~~~~~~
+
+.. method:: alias(*args, **kwargs)
```
You we really allow `*args` to be passed here? It doesn't look like its tested below and it seems like using `alias` with the few number of expressions that implement default naming is unlikely.

alextatarinov commented:
@charettes Thanks for the review! There actually was a bug - using `.alias()` cleared all previously added annotations. Added tests, made suggested changes, and rebased on master. I've also found that the `filter` in the `test_filter_alias_with_f` was producing an empty `QuerySet`. This test was copied from `test_filter_annotation_with_f`, so I fixed that as well (I can revert that change if it is inappropriate and needs a separate pull request).
I think we should keep the API consistent with `annotate`, so I've added a test for `args` support.

charettes commented on tests/annotations/tests.py:
```diff
@@ -119,17 +119,79 @@ def test_basic_annotation(self):
+            is_book_alias=Value(1)
+        ).annotate(is_book=F('is_book_alias'))
+        self.assertFalse(hasattr(books.first(), 'is_book_alias'))
+        for book in books:
+            self.assertEqual(book.is_book, 1)
```
Either use `subTest` or collect all the attributes and use `assertCountEqual`.

charettes commented on tests/annotations/tests.py:
```diff
@@ -119,17 +119,79 @@ def test_basic_annotation(self):
+        books = Book.objects.alias(
+            is_book=Value(1)
+        ).annotate(is_book=F('is_book'))
+        for book in books:
+            self.assertEqual(book.is_book, 1)
```
Ditto

charettes commented on tests/annotations/tests.py:
```diff
@@ -119,17 +119,79 @@ def test_basic_annotation(self):
+            is_book_alias=Value(1)
+        ).annotate(is_book=Coalesce('is_book_alias', 0))
+        self.assertFalse(hasattr(books.first(), 'is_book_alias'))
+        for book in books:
+            self.assertEqual(book.is_book, 1)
```
Ditto

charettes commented on tests/annotations/tests.py:
```diff
@@ -119,17 +119,79 @@ def test_basic_annotation(self):
+            another_rating_alias=F('rating')
+        ).annotate(another_rating=F('another_rating_alias'))
+        self.assertFalse(hasattr(books.first(), 'another_rating_alias'))
+        for book in books:
+            self.assertEqual(book.another_rating, book.rating)
```
Ditto

charettes commented on tests/annotations/tests.py:
```diff
@@ -119,17 +119,79 @@ def test_basic_annotation(self):
+            num_awards_alias=F('publisher__num_awards')
+        ).annotate(num_awards=F('num_awards_alias'))
+        self.assertFalse(hasattr(books.first(), 'num_awards_alias'))
+        for book in books:
+            self.assertEqual(book.num_awards, book.publisher.num_awards)
```
Ditto

charettes commented on tests/annotations/tests.py:
```diff
@@ -167,6 +229,21 @@ def test_annotate_with_aggregation(self):
+        book = books.first()
+        self.assertFalse(hasattr(book, 'is_book_alias'))
+        self.assertFalse(hasattr(book, 'rating_count_alias'))
+        for book in books:
+            self.assertEqual(book.is_book, 1)
+            self.assertEqual(book.rating_count, 1)
```
Ditto

charettes commented on tests/annotations/tests.py:
```diff
@@ -188,6 +265,15 @@ def test_aggregate_over_annotation(self):
+    def test_aggregate_over_alias(self):
+        with self.assertRaisesMessage(
+            FieldError,
+            'Cannot aggregate over the alias. Use annotate() to promote the alias'
```
```suggestion
            "Cannot aggregate over the 'other_age' alias. Use annotate() to promote it"
```

charettes commented on tests/annotations/tests.py:
```diff
@@ -227,6 +313,11 @@ def test_distinct_on_with_annotation(self):
 
+    @skipUnlessDBFeature('can_distinct_on_fields')
+    def test_distinct_on_with_alias(self):
+        with self.assertRaisesMessage(FieldError, "Cannot resolve keyword 'other_rating' into field."):
```
Should we also have a pointer here of the form

```suggestion
        with self.assertRaisesMessage(FieldError, "Cannot distinct on 'other_rating' alias. Use annotate to promote it"):
```

alextatarinov replied:
I am not sure where this logic should go and what to do with other methods like `values`. 
The idea was to rely on existing code to error as if the field does not exist when trying to use the alias, but the aggregation failed with some cryptic message so I've added the correct error message.

felixxm replied:
I improved messages for `values()` and `values_list()`. `distinct()` is more complicated because the exception is raised by `names_to_path()` that is used in many places, IMO we can leave the current message.

charettes commented on django/db/models/sql/query.py:
```diff
@@ -1707,6 +1710,8 @@ def resolve_ref(self, name, allow_joins=True, reuse=None, summarize=False):
                 # aggregate() elements reference an existing annotation. In
                 # that case we need to return a Ref to the subquery's annotation.
+                if name not in self.annotation_select:
+                    raise FieldError('Cannot aggregate over the alias. Use annotate() to promote the alias')
```
```suggestion
                    raise FieldError(f"Cannot aggregate over the '{name}' alias. Use annotate() to promote it")
```

alextatarinov commented on docs/releases/3.2.txt:
```diff
@@ -52,6 +52,16 @@ With automatic ``AppConfig`` discovery, ``default_app_config`` is no longer
+later reuse with other ``QuerySet`` methods via a new
+:meth:`~django.db.models.query.QuerySet.alias` method. This is useful when
+the result of the expression itself is not needed but is used for filtering,
+ordering, or as a part of a complex expression. By not selecting the actual value, one can construct
+a smaller SQL query, return less data from the database, and avoid double evaluation
+of costly expressions, which results in better performance.
```
Much of the motivation for me and the original issue author was the query execution time, and query construction time by the ORM. When using `alias()` with `filter()`, the expression won't be included both in the SELECT and WHERE clauses as with `annotate()`. Also, it is not rare to make some annotations that may or may not be used for filtering and ordering later, but most of them are discarded, so making them `alias` will improve performance.
I am not sure whether any backend will detect duplicated expressions and cache the results, but it seems that avoiding duplicated expressions leads to a significant performance boost, especially when annotations are `Subquery`.
So, I believe that performance boost due to the fewer data returned is not the most important one, and we should emphasize on the query execution time.

> Doing a full sequential scan (no index) with both calculating the to_tsvector() for return and in query predicate is 27,000ms without recalculating the to_tsvector for return, 46,700ms with recalculating it.
> After having created a functional index on the to_tsvector() call, and using a relatively common search term, the time to return without recalculating the to_tsvector was 1.7ms, vs 21,200ms with recalculating the ts_vector.

charettes replied:
> query construction time by the ORM.

I'm not sure this is actually significant, we're talking about a single `as_sql` call removal here since the expression will be resolved once in both cases. I mean, sure that's one less expression compilation down the line but I don't think it's significant enough to warrant a mention in the docs.

Speaking of which we should probably add a test to make sure that `JOIN` added for an unused alias is pruned 

e.g. `Book.objects.alias(author_name=Lower('author__name'))` should not create a `JOIN`.

> Also, it is not rare to make some annotations that may or may not be used for filtering and ordering later, but most of them are discarded, so making them alias will improve performance.

I agree that it isn't rare but I wouldn't say it's common either. From experience I can count the number of times this has come up on the past projects I worked on a single hand.

> I am not sure whether any backend will detect duplicated expressions and cache the results, but it seems that avoiding duplicated expressions leads to a significant performance boost, especially when annotations are Subquery.

While this is mentioned on the ticket this is now easy to work around by passing `Subquery` and other such expressions directly to `filter` or `order_by` which should cover 90% of the cases thanks to ticket-25367 being fixed.

> So, I believe that performance boost due to the fewer data returned is not the most important one, and we should emphasize on the query execution time.

I agree that the above statement should be adapted to reflect this. What about

> Not selecting the actual value can allow the database to perform some optimization and result in less data being returned. This can result in better performance when dealing with certain type of columns or complex expressions.

alextatarinov replied:
I think I will need to come up with some performance tests for query construction to make a final decision, but I've experienced significant CPU time cost due to the query size. 

> Speaking of which we should probably add a test to make sure that JOIN added for an unused alias is pruned

As stated in the ticket, 
> all other logic remains the same: despite the expression is not selected, the joins and group by are the same as for annotation. 

However, it seems like LEFT OUTER joins with FK can be pruned, I'll take a look and the current behavior.

About the next two points: when the complex expression is constructed by parts, in different places and only later somewhere may be used for filtering or sorting, most of the annotations are useless, because only final expression is valuable, and that expression can be used only in filter clause, not required to select. That's basically the whole point of the ticket, to produce `SELECT x, y FROM z WHERE (SubExpression1 + SubExpression2 + SubExpression3)  = FOO` instead of 
`SELECT x, y, SubExpression1 as a, SubExpression2 as b, SubExpression3 as c, (SubExpression1 + SubExpression2 + SubExpression3) as d  FROM z WHERE (SubExpression1 + SubExpression2 + SubExpression3) = FOO`
where each SubExpression may be a `Subquery` or another costly expression.

For me, the ticket-25367 doesn't cover most of the usages, because often you make annotations and filtering in completely different parts of the codebase, i.e: annotate in `QuerySet`, filter in the view.

I think it is hard to explain most use cases and benefits, we may leave that to users and blog posts authors. But I believe that the query without expressions in the SELECT will perform better, refer to the ticket for some examples of 1.7ms vs 21200ms due to SELECT. What about

> Not selecting the unused value removes redundant work from the database which leads to better performance.

charettes replied:
> That's basically the whole point of the ticket ...
> For me, the ticket-25367 doesn't cover most of the usages, because often you make annotations and filtering in completely different parts of the codebase, i.e: annotate in QuerySet, filter in the view.

I would argue that's bad practice to add unused annotations at the queryset/CBV level if it's only going to be used for filtering or ordering purpose under certain circumstances. I suspect folks who are investing time identifying the origin of some slow queries generated by the ORM are already avoiding this pattern.

> What about
>> Not selecting the unused value removes redundant work from the database which should result in better performance.

That works for me.

alextatarinov replied:
About the JOIN - it is not trimmed, and I'm very unsure it can be done without significant effort. From the beginning, I've decided this is actually an intended behavior, so I suggest to keep it as is unless someone can propose a simple solution without caveats.

charettes replied:
> From the beginning, I've decided this is actually an intended behavior, so I suggest to keep it as is unless someone can propose a simple solution without caveats.

Makes sense.

felixxm commented on docs/releases/3.2.txt:
```diff
@@ -53,6 +53,16 @@ needed. As a consequence, it's deprecated.
 
 See :ref:`configuring-applications-ref` for full details.
 
+Expressions alias
+-----------------
+
+Expressions that don't need to be selected can now be saved as an alias for
+later reuse with other ``QuerySet`` methods via a new
+:meth:`~django.db.models.query.QuerySet.alias` method. This is useful when
+the result of the expression itself is not needed but is used for filtering,
+ordering, or as a part of a complex expression. Not selecting the unused value
+removes redundant work from the database which should result in better performance.
+
```
It's not a major feature, IMO, I would move it to the `Minor features -> Models`, maybe:
```
* The new :meth:`.QuerySet.alias` method allows creating reusable aliases for
  expressions that don't need to be selected but are useful for filtering,
  ordering, or as a part of a complex expression.
```

alextatarinov replied:
I would argue that adding a new public `QuerySet` method can be considered a major feature. As far as I can understand, minor features are generally amendments to existing functionality like behavior change or introducing new parameters. But obviously I don't have enough experience to make a decision. What would you say?

felixxm replied:
> minor features are generally amendments to existing functionality like behavior change or introducing new parameters

Not really, e.g. [`QuerySet.bulk_update()`](https://docs.djangoproject.com/en/3.0/releases/2.2/#models) or almost all new database functions are in minor features. We can decide to promote it before the 3.2 release but for now I would add it to the minor features.

felixxm commented on docs/ref/models/querysets.txt:
```diff
@@ -268,6 +268,46 @@ control the name of the annotation::
+
+.. versionadded:: 3.2
+
+Same as :meth:`~django.db.models.query.QuerySet.annotate()`, but instead of
+annotating objects in the ``QuerySet``, saves the expression for later reuse with
+other ``QuerySet`` methods. This is useful when
+the result of the expression itself is not needed but is used for filtering,
+ordering, or as a part of a complex expression. Not selecting the unused value
+removes redundant work from the database which should result in better performance.
+
+For example, if you were manipulating a list of blogs, you may want
+to find blogs with more than 5 entries, but not actually interested
+in the exact number of entries::
+
+    >>> from django.db.models import Count
+    >>> blogs = Blog.objects.alias(entries=Count('entry')).filter(entries__gt=5)
+
+``alias()`` can be used in conjunction with
+:meth:`~django.db.models.query.QuerySet.annotate()`,
+:meth:`~django.db.models.query.QuerySet.filter()`,
+:meth:`~django.db.models.query.QuerySet.order_by()`
+and :meth:`~django.db.models.query.QuerySet.update()`.
+To use aliased expression with other methods (e.g. aggregate), one must promote alias to annotation:
+
+    >>> Blog.objects.alias(entries=Count('entry')).annotate(entries=F('entries')).aggregate(Sum('entries'))
+
+:meth:`~django.db.models.query.QuerySet.filter()` and
+:meth:`~django.db.models.query.QuerySet.order_by()`
+can take expressions directly, but expression construction and usage often does
+not happen in the same place (for example, ``QuerySet``
+method creates expression, to be later used by the view).
+Also ``alias()`` allows you to build complex expressions incrementally,
+possibly spanning multiple methods and modules, refer to the expression parts
+by their aliases and only use
+:meth:`~django.db.models.query.QuerySet.annotate()` for the final result.
```
Wrap docs at 79 chars. You can also simplify all references to `QuerySet`\'s methods in this doc, e.g. `` :meth:`~django.db.models.query.QuerySet.filter()` `` -> `` :meth:`filter` ``. Also `()` are added automatically when you use `:meth:`.

felixxm commented on docs/ref/models/querysets.txt:
```diff
@@ -268,6 +268,46 @@ control the name of the annotation::
+:meth:`~django.db.models.query.QuerySet.filter()`,
+:meth:`~django.db.models.query.QuerySet.order_by()`
+and :meth:`~django.db.models.query.QuerySet.update()`.
+To use aliased expression with other methods (e.g. aggregate), one must promote alias to annotation:
```
```suggestion
To use aliased expression with other methods (e.g. :meth:`aggregate`), one must promote alias to annotation:
```

felixxm commented on docs/ref/models/querysets.txt:
```diff
@@ -268,6 +268,46 @@ control the name of the annotation::
+:meth:`~django.db.models.query.QuerySet.order_by()`
+can take expressions directly, but expression construction and usage often does
+not happen in the same place (for example, ``QuerySet``
+method creates expression, to be later used by the view).
```
```suggestion
method creates expressions, for later use in views).
```

felixxm commented on docs/ref/models/querysets.txt:
```diff
@@ -268,6 +268,46 @@ control the name of the annotation::
+can take expressions directly, but expression construction and usage often does
+not happen in the same place (for example, ``QuerySet``
+method creates expression, to be later used by the view).
+Also ``alias()`` allows you to build complex expressions incrementally,
```
```suggestion
``alias()`` allows building complex expressions incrementally,
```

felixxm commented on tests/annotations/tests.py:
```diff
@@ -119,17 +119,84 @@ def test_basic_annotation(self):
         for book in books:
             self.assertEqual(book.is_book, 1)
 
+    def test_basic_alias(self):
```
I think we should move all new tests to a separate class, e.g. `AliasTests`.

alextatarinov replied:
@felixxm Does it need a separate module or can be kept in `annotations/tests.py`?

felixxm replied:
We can keep it in the same file.

felixxm commented on tests/annotations/tests.py:
```diff
@@ -188,6 +271,15 @@ def test_aggregate_over_annotation(self):
         self.assertEqual(agg['otherage_sum'], other_agg['age_sum'])
 
+    def test_aggregate_over_alias(self):
+        with self.assertRaisesMessage(
+            FieldError,
+            "Cannot aggregate over the 'other_age' alias. Use annotate() to promote it"
+        ):
```
```suggestion
        msg = (
            "Cannot aggregate over the 'other_age' alias. Use annotate() to "
            "promote it."
        )
        with self.assertRaisesMessage(FieldError, msg):
```
Use the same pattern in other `assertRaisesMessage`.

felixxm commented on tests/annotations/tests.py:
```diff
@@ -235,9 +332,19 @@ def test_filter_annotation(self):
     def test_filter_annotation_with_f(self):
         books = Book.objects.annotate(
             other_rating=F('rating')
-        ).filter(other_rating=3.5)
```
Please revert the unrelated change to an existing test.

felixxm commented on tests/annotations/tests.py:
```diff
@@ -361,7 +492,7 @@ def test_defer_annotation(self):
 
         with self.assertRaisesMessage(FieldDoesNotExist, "Book has no field named 'other_rating'"):
-            book = qs.defer('other_rating').get(other_rating=4)
+            qs.defer('other_rating').get(other_rating=4)
```
This is also unrelated, please revert it.

felixxm commented on tests/annotations/tests.py:
```diff
@@ -419,6 +560,16 @@ def test_order_by_aggregate(self):
+        authors = Author.objects.values('age').alias(age_count=Count('age')).order_by('age_count', 'age')
+        self.assertFalse(hasattr(authors.first(), 'age_count'))
+        self.assertQuerysetEqual(
+            authors, [
+                (25,), (34,), (35,), (37,), (45,), (46,), (57,), (29,),
+            ],
+            lambda a: (a['age'],)
```
```suggestion
            authors,
            [25, 34, 35, 37, 45, 46, 57, 29],
            lambda a: a['age']
```

felixxm commented on tests/basic/tests.py:
```diff
@@ -574,6 +574,7 @@ class ManagerTest(SimpleTestCase):
+        'alias',
         'complex_filter',
         'exclude',
         'in_bulk',
```
We could also add `alias` to the `queries.test_qs_combinators.QuerySetSetOperationTests.test_unsupported_operations_on_combined_qs()`.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
@alextatarinov Thanks :+1: 

I pushed small edits to docs, simplified tests, improved error messages for `values()`/`values_list()`, and added tests for `dates()`, `defer()`, and `only().`
